core/pax-exam/src/main/java/org/ops4j/pax/exam/ExamJavaRunner.java

```diff
@@ -134,13 +136,59 @@ public synchronized void exec(final String[] vmOptions,
-        HashSet<String> envSet = new HashSet<String>(Arrays.asList(envOptions));
-        Map<String, String> env = pb.environment();
-        HashSet<String> toRemove = new HashSet<String>(env.keySet());
-        toRemove.removeAll(envSet);
-        for (String key : toRemove) {
-            env.remove(key);
+        final Map<String, String> envMap = toKeyValueMap(envOptions);
+
+        final Map<String, String> env = pb.environment();
+
+        // Remove all items from the ProcessBuilder's environment,
+        // that are not explicitly referenced in the envMap's keyset
+        final Set<String> toRemove = new HashSet<String>(env.keySet());
+        toRemove.removeAll(envMap.keySet());
+        env.keySet().removeAll(toRemove);
+
+        // Any supplied values should be applied, possibly overriding an initial value from the ProcessBuilder.
+        envMap.entrySet().stream()
+                .filter(entry -> entry.getValue() != null)
+                .forEach(entry -> env.put(entry.getKey(), entry.getValue()));
+    }
+
+    /**
+     * Converts an array of strings in the format "key=value" into a map of key-value pairs.
+     * If a string does not contain an '=' character, the entire string is treated as the key,
+     * and the corresponding value is set to {@code null}.
+     *
+     * @param envOptions an array of strings, each representing an environment option in the
+     *                   format "key=value". Strings without an '=' will have {@code null} as the value.
+     * @return a {@code Map<String, String>} where each key-value pair is derived from the input array.
+     *         Keys are extracted from the portion of the string before the first '=' character,
+     *         and values are extracted from the portion after the '=' character. If no '=' is
+     *         present, the value is set to {@code null}.
+     */
+    private static Map<String, String> toKeyValueMap(String[] envOptions) {
+        final Map<String, String> envMap = new HashMap<>();
+        for (String envOption : envOptions) {
+            if (envOption == null || envOption.isEmpty()) {
+                throw new IllegalArgumentException("Null or empty entry in envOptions");
+            }
+            final int equalsPosition = envOption.indexOf('=');
+            final String key;
+            final String value;
+            if (equalsPosition < 0) {
+                key = envOption.trim();
+                value = null;
+            } else {
+                key = envOption.substring(0, equalsPosition).trim();
+                value = envOption.substring(equalsPosition + 1).trim();
+            }
+            if (key.isEmpty()) {
+                throw new IllegalArgumentException("Input " + envOption + " resulted in an empty key");
+            }
+            if (envMap.containsKey(key)) {
+                throw new IllegalArgumentException("Duplicate key found: " + key);
+            }
```

Comment:
Did you encounter problems in above situations where `IllegalArgumentException`s are thrown? Or would logging a warning be sufficient?

Reply:
No I did not encounter issues, I added this as sanity check on the user input. I guess it personal preference to prefer fail fast by throwing an exception, as in my experience warnings are often overlooked. 
Logging a warning is fine by me as well, please let me know if you prefer that.